_posts/2016-10-26-jwt-artsy-journey.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+title: "JSON Web Tokens: Artsy's Journey"
+date: 2016-10-26 11:03
+comments: true
+author: ashkan
+categories: [Authentication, JWT, API]
+---
+
+At Artsy we currently have ~1730 client applications hitting our API and requesting authentication. When a user successfully authenticates through one of these clients, we want to embed basic user/application data in the resulting token rather than have to look up a session ID in the database on each request. For that we want to use JWT.
+
+JWT (JSON Web Token) is a self-contained, secure and standard way of transmitting data between applications and clients as JSON objects.
+<!-- more -->
+JWT has three separate sections which are separated by `.`: header, payload and signature.
+
+`<header>.<payload>.<signature>`
+
+Once decoded:
+
+- The header section contains JSON defining the type of token and how to treat its contents.
+
+- The payload seciton includes _claims_ and data provided by token issuer about the user/application, things like token expiration time, user id and etc.
+
+- Signature section is used to verify the sender of the JWT and make sure token wasn't modified along the way.
+
+<a href="https://jwt.io/introduction/" target="_blank">JWT</a> and <a href="https://tools.ietf.org/html/rfc7519" target="_blank">RFC 7519</a> provide more details about each section of JWT and how encoding, decoding and verification works.
+
+
+# Where we were
+
+After a successful login, our API would generate a custom JSON object including some basic information about the user and application, then encrypt it using a single server-side secret.
+For every authenticated request we would have to decrypt the token, make sure user and application are valid and application still has access to our API. We are already _stateless_ since we don't store tokens in our database.
+In this manner, we avoided needing to store sessions in the database. However, client applications had to make an API call to get any information about the authenticated user.
```

Comment:
_We are already stateless..._ and _...we avoid needing to store..._ seem redundant (probably my fault). You can consolidate or just remove the 2nd.